x/cron/cron_test.go

```diff
@@ -0,0 +1,344 @@
+				errs: map[string]error{
+					"test/1": errors.ErrState,
+				},
+			},
+		},
+		"a mixture of tasks, some are due and some are successful": {
+			Tasks: []*task{
+				{
+					RunAt:    now.Add(time.Hour),
+					Auth:     nil,
+					Msg:      weavetest.Msg{RoutePath: "wait/1"},
+					WantExec: false,
+				},
+				{
+					RunAt:           now.Add(-time.Hour),
+					Auth:            nil,
+					Msg:             weavetest.Msg{RoutePath: "due/success"},
+					WantExec:        true,
+					WantExecSuccess: true,
+				},
+				{
+					RunAt:           now.Add(-time.Hour),
+					Auth:            nil,
+					Msg:             weavetest.Msg{RoutePath: "due/failure"},
+					WantExec:        true,
+					WantExecSuccess: false,
+				},
+				{
+					RunAt:    now.Add(time.Hour),
+					Auth:     nil,
+					Msg:      weavetest.Msg{RoutePath: "wait/2"},
+					WantExec: false,
+				},
+			},
+			WantTickerErr: nil,
+			Handler: cronHandler{
+				errs: map[string]error{
+					"due/success": nil,
+					"due/failure": errors.ErrHuman,
+				},
+			},
+		},
+	}
+
+	for testName, tc := range cases {
+		t.Run(testName, func(t *testing.T) {
+			db := store.MemStore()
+			migration.MustInitPkg(db, "cron")
+			enc := NewTestTaskMarshaler(&weavetest.Msg{})
+			scheduler := NewScheduler(enc)
+			ticker := NewTicker(&tc.Handler, enc)
+
+			for i, task := range tc.Tasks {
+				tid, err := scheduler.Schedule(db, task.RunAt, task.Auth, &task.Msg)
+				if err != nil {
+					t.Fatalf("cannot schedule #%d (%q) task: %s", i, task.Msg.Path(), err)
+				}
+				task.id = tid
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+			ctx = weave.WithBlockTime(ctx, now)
+
+			// Use tick instead of Tick method so that the error is
+			// returned instead of terminating the process.
+			// We need to tick only once in order to process all
+			// tasks that are due.
+			tags, _, err := ticker.tick(ctx, db)
+			if !tc.WantTickerErr.Is(err) {
+				t.Fatalf("unexpected ticker error: %+v", err)
+			}
+
+			for i, task := range tc.Tasks {
+				wasExec := containsPairValue(tags, task.id)
+				if wasExec != task.WantExec {
+					t.Fatalf("task #%d (%q) unexpected execution state: %v", i, task.Msg.Path(), wasExec)
+				}
+			}
+
+			// For every executed task, there must be a state information persisted.
+			b := NewTaskResultBucket()
+			for i, task := range tc.Tasks {
+				if !task.WantExec {
+					continue
+				}
+				var tr TaskResult
+				if err := b.One(db, task.id, &tr); err != nil {
+					t.Fatalf("expected #%d (%q) task result, got %s", i, task.Msg.Path(), err)
+				}
+				if tr.Successful != task.WantExecSuccess {
+					t.Fatalf("exected task #%d (%q) to be success=%v: %q", i, task.Msg.Path(), task.WantExecSuccess, tr.Info)
+				}
+			}
+		})
+	}
+}
+
+func containsPairValue(pairs []common.KVPair, item []byte) bool {
+	for _, p := range pairs {
+		if bytes.Equal(p.Value, item) {
+			return true
+		}
+	}
+	return false
+}
+
```

Comment:
req: please test Scheduler.Delete

Reply:
Done in https://github.com/iov-one/weave/pull/850/commits/5b1ca21e5a8c41051e044a9303ac1c552cce283e